File: dispatcher/src/queue.ts
import amqp from 'amqplib';

export class QueueClient {
  private conn?: amqp.Connection;
  private channel?: amqp.Channel;

  constructor(private url: string) {}

  async connect(): Promise<void> {
    this.conn = await amqp.connect(this.url);
    this.channel = await this.conn.createChannel();
  }

  async consume(queue: string, onMessage: (msg: amqp.ConsumeMessage) => Promise<void>): Promise<void> {
    if (!this.channel) throw new Error('Channel not initialized. Run connect() first.');
    
    await this.channel.assertQueue(queue, { durable: true });
    await this.channel.prefetch(10); // Limit unacknowledged messages per worker

    await this.channel.consume(queue, async (msg) => {
      if (msg) {
        try {
          await onMessage(msg);
          this.channel?.ack(msg);
        } catch (err) {
          console.error('Error handling message:', err);
          // Worker code handles manual nack/ack, this is a fallback
          this.channel?.nack(msg, false, true);
        }
      }
    });
  }

  async publishToExchange(exchange: string, routingKey: string, content: Buffer): Promise<void> {
    if (!this.channel) throw new Error('Channel not initialized.');
    await this.channel.publish(exchange, routingKey, content, { persistent: true });
  }

  async close(): Promise<void> {
    await this.channel?.close();
    await this.conn?.close();
  }
}

File: gateway/internal/queue/queue.go
package queue

import (
	"context"
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

type RabbitPublisher struct {
	conn    *amqp.Connection
	channel *amqp.Channel
}

func NewRabbitPublisher(url string) (*RabbitPublisher, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, err
	}

	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, err
	}

	return &RabbitPublisher{
		conn:    conn,
		channel: ch,
	}, nil
}

func (p *RabbitPublisher) Publish(ctx context.Context, exchange, routingKey string, body []byte) error {
	log.Printf("Publishing message to exchange %s with routing key %s", exchange, routingKey)
	return p.channel.PublishWithContext(ctx,
		exchange,
		routingKey,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType: "application/json",
			Body:        body,
		},
	)
}

func (p *RabbitPublisher) Close() {
	if p.channel != nil {
		p.channel.Close()
	}
	if p.conn != nil {
		p.conn.Close()
	}
}
